pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -559,6 +588,147 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+          tableNameWithType);
+      LOGGER.error(errMsg, e);
+    }
+    return tableSizePerReplicaInBytes;
+  }
+
+  private RebalanceSummaryResult calculateDryRunSummary(Map<String, Map<String, String>> currentAssignment,
+      Map<String, Map<String, String>> targetAssignment, String tableNameWithType, String rebalanceJobId) {
+    LOGGER.info("Calculating rebalance summary for table: {} with rebalanceJobId: {}",
+        tableNameWithType, rebalanceJobId);
+    int existingReplicationFactor = 0;
+    int newReplicationFactor = 0;
+    Map<String, Set<String>> existingServersToSegmentMap = new HashMap<>();
+    Map<String, Set<String>> newServersToSegmentMap = new HashMap<>();
+
+    for (Map.Entry<String, Map<String, String>> entrySet : currentAssignment.entrySet()) {
+      existingReplicationFactor = entrySet.getValue().size();
+      for (Map.Entry<String, String> segmentEntrySet : entrySet.getValue().entrySet()) {
+        existingServersToSegmentMap.putIfAbsent(segmentEntrySet.getKey(), new HashSet<>());
+        existingServersToSegmentMap.get(segmentEntrySet.getKey()).add(entrySet.getKey());
+      }
+    }
+
+    for (Map.Entry<String, Map<String, String>> entrySet : targetAssignment.entrySet()) {
+      newReplicationFactor = entrySet.getValue().size();
+      for (Map.Entry<String, String> segmentEntrySet : entrySet.getValue().entrySet()) {
+        newServersToSegmentMap.putIfAbsent(segmentEntrySet.getKey(), new HashSet<>());
+        newServersToSegmentMap.get(segmentEntrySet.getKey()).add(entrySet.getKey());
+      }
+    }
+    RebalanceSummaryResult.RebalanceChangeInfo replicationFactor
+        = new RebalanceSummaryResult.RebalanceChangeInfo(existingReplicationFactor, newReplicationFactor);
+
+    int existingNumServers = existingServersToSegmentMap.keySet().size();
+    int newNumServers = newServersToSegmentMap.keySet().size();
+    RebalanceSummaryResult.RebalanceChangeInfo numServers
+        = new RebalanceSummaryResult.RebalanceChangeInfo(existingNumServers, newNumServers);
+
+    List<InstanceConfig> instanceConfigs = _helixDataAccessor.getChildValues(
+        _helixDataAccessor.keyBuilder().instanceConfigs(), true);
+    Map<String, List<String>> instanceToTagsMap = new HashMap<>();
+    for (InstanceConfig instanceConfig : instanceConfigs) {
+      instanceToTagsMap.put(instanceConfig.getInstanceName(), instanceConfig.getTags());
+    }
+
+    Map<String, RebalanceSummaryResult.ServerSegmentChangeInfo> serverSegmentChangeInfoMap = new HashMap<>();
+    int segmentsNotMoved = 0;
+    int numServersGettingDataAdded = 0;
+    for (Map.Entry<String, Set<String>> entry : newServersToSegmentMap.entrySet()) {
+      String server = entry.getKey();
+      Set<String> segmentMap = entry.getValue();
+      int totalNewSegments = segmentMap.size();
+
+      Set<String> newSegmentList = new HashSet<>(segmentMap);
+      Set<String> existingSegmentList = new HashSet<>();
```

Comment:
nit: perhaps remove `List` suffix from the var name if using Set<> for bit less confusion

Reply:
done